src/Integrations/Integrations/Swoole/SwooleIntegration.php

```diff
@@ -0,0 +1,202 @@
+                    if (\array_key_exists($normalizedHeader, $allowedHeaders)) {
+                        $rootSpan->meta["http.request.headers.$normalizedHeader"] = $value;
+                    }
+                }
+                consume_distributed_tracing_headers(function ($key) use ($headers) {
+                    return $headers[$key] ?? null;
+                });
+
+                if (\dd_trace_env_config("DD_TRACE_CLIENT_IP_ENABLED")) {
+                    $res = extract_ip_from_headers($headers + ['REMOTE_ADDR' => $request->server['remote_addr']]);
+                    $rootSpan->meta += $res;
+                }
+
+                if (isset($headers["user-agent"])) {
+                    $rootSpan->meta["http.useragent"] = $headers["user-agent"];
+                }
+
+                $rawContent = $request->rawContent();
+                if ($rawContent) {
+                    // The raw content will always be populated if the request is a POST request, independent of the
+                    // Content-Type header.
+                    // However, it may not be json-decodable
+                    $postFields = json_decode($rawContent, true);
+                    if (is_null($postFields)) {
+                        // Fallback to the post fields, which is an array
+                        // This array is not always populated, depending on the Content-Type header
+                        $postFields = $request->post;
+                    }
+                }
+                if (!empty($postFields)) {
+                    $postFields = Normalizer::sanitizePostFields($postFields);
+                    foreach ($postFields as $key => $value) {
+                        $rootSpan->meta["http.request.post.$key"] = $value;
+                    }
+                }
+
+                $normalizedPath = Normalizer::uriNormalizeincomingPath(
+                    $request->server['request_uri']
+                    ?? $request->server['path_info']
+                    ?? '/'
+                );
+                $rootSpan->resource = $request->server['request_method'] . ' ' . $normalizedPath;
+                $rootSpan->meta[Tag::HTTP_METHOD] = $request->server['request_method'];
+
+                $serverProtocol = $request->server['server_protocol'] ?? 'HTTP/1.1';
+                $scheme = strpos($serverProtocol, 'HTTPS') !== false ? 'https://' : 'http://';
+                $host = $headers['host'] ?? ($request->server['remote_addr'] . ':' . $request->server['server_port']);
+                $path = $request->server['request_uri'] ?? $request->server['path_info'] ?? '';
+                $query = isset($request->server['query_string']) ? '?' . $request->server['query_string'] : '';
+                $url = $scheme . $host . $path . $query;
+                $rootSpan->meta[Tag::HTTP_URL] = Normalizer::uriNormalizeincomingPath($url);
+            },
+            function (HookData $hook) use (&$rootSpan, $integration) {
```

Comment:
This doesn't look right to me. &$rootSpan will not necessarily be the root span the Request was started with - multiple requests may run in parallel with swoole.

In the begin hook, use `$rootSpan = $hook->span(new \DDTrace\SpanStack);`. Then you also don't have to manually close the span in the end hook.

Reply:
Changed :+1: That makes me wonder whether we should do the same for the RoadRunner integration 🤔

Reply:
Makes sense there too, but for roadrunner it's not a bug at least.